src/wallet.h

```diff
@@ -930,16 +944,38 @@ class CWalletTx : public CMerkleTx
 
     std::set<uint256> GetConflicts() const;
 
+
 private:
-    void FillValuesAndBlindingFactors() const {
+    void FillValuesAndBlindingFactors(unsigned int nOut) const {
         if (!vAmountsOut.size()) {
             vAmountsOut.resize(vout.size());
             vBlindingFactors.resize(vout.size());
-            for (unsigned int i = 0; i < vout.size(); i++) {
-                std::vector<unsigned char> nonce(32, 0);
-                int res = UnblindOutput(pwallet->blinding_key, vout[i], vAmountsOut[i], vBlindingFactors[i]);
-                if (!res)
-                    vAmountsOut[i] = -1;
+            vBlindingKeys.resize(vout.size());
+        }
+        if (vAmountsOut[nOut] == 0) {
+            std::vector<unsigned char> nonce(32, 0);
+            int res = 0;
+            CKey blinding_key;
+            if (!res) {
+                // For unblinded outputs.
+                res = UnblindOutput(blinding_key, vout[nOut], vAmountsOut[nOut], vBlindingFactors[nOut]);
+            }
+            if (!res && (blinding_key = pwallet->GetBlindingKey(&vout[nOut].scriptPubKey)).IsValid()) {
+                // For outputs using derived blinding.
+                res = UnblindOutput(blinding_key, vout[nOut], vAmountsOut[nOut], vBlindingFactors[nOut]);
+                if (res) {
+                    vBlindingKeys[nOut] = blinding_key.GetPubKey();
+                }
+            }
+            if (!res && (blinding_key = pwallet->GetBlindingKey(NULL)).IsValid()) {
```

Comment:
We're having issues with none of these three cases firing with either unblinded or static blinded outputs. We get:

alphad: base58.cpp:240: CBitcoinAddress& CBitcoinAddress::AddBlindingKey(const CPubKey&): Assertion `pubkey.size() == 33' failed.